File: services/price-service/internal/adapters/base/chain.go
package base

import (
	"fmt"
	"io"
	"net/http"
	"net/url"
	"regexp"
	"strings"
	"time"

	"github.com/kosarica/price-service/internal/adapters/config"
	httpclient "github.com/kosarica/price-service/internal/http"
	"github.com/kosarica/price-service/internal/http/ratelimit"
	"github.com/kosarica/price-service/internal/types"
)

// BaseAdapterConfig contains configuration for base chain adapter
type BaseAdapterConfig struct {
	Slug                   string
	Name                   string
	SupportedTypes         []types.FileType
	ChainConfig            config.ChainConfig
	FilenamePrefixPatterns []string
	FileExtensionPattern   *regexp.Regexp
	RateLimitOverrides     *ratelimit.PartialConfig
}

// ChainAdapter interface defines the contract for all chain adapters
type ChainAdapter interface {
	Slug() string
	Name() string
	SupportedTypes() []types.FileType
	Discover(targetDate string) ([]types.DiscoveredFile, error)
	Fetch(file types.DiscoveredFile) (*types.FetchedFile, error)
	Parse(content []byte, filename string, options *types.ParseOptions) (*types.ParseResult, error)
	ExtractStoreIdentifier(file types.DiscoveredFile) *types.StoreIdentifier
	ValidateRow(row types.NormalizedRow) types.NormalizedRowValidation
	ExtractStoreMetadata(file types.DiscoveredFile) *types.StoreMetadata
}

// BaseChainAdapter provides common implementations for all chain adapters
type BaseChainAdapter struct {
	slug                   string
	name                   string
	supportedTypes         []types.FileType
	config                 config.ChainConfig
	filenamePrefixPatterns []*regexp.Regexp
	fileExtensionPattern   *regexp.Regexp
	rateLimiter            *ratelimit.RateLimiter
	rateLimitConfig        ratelimit.Config
	httpClient             *httpclient.Client
}

// NewBaseChainAdapter creates a new base chain adapter
func NewBaseChainAdapter(cfg BaseAdapterConfig) (*BaseChainAdapter, error) {
	// Validate supported types is not empty
	if len(cfg.SupportedTypes) == 0 {
		return nil, fmt.Errorf("%s: SupportedTypes cannot be empty", cfg.Slug)
	}

	fileExtensionPattern := cfg.FileExtensionPattern
	if fileExtensionPattern == nil {
		fileExtensionPattern = regexp.MustCompile(`\.(csv|CSV)$`)
	}

	filenamePrefixPatterns := make([]*regexp.Regexp, 0, len(cfg.FilenamePrefixPatterns)+2)
	for _, pattern := range cfg.FilenamePrefixPatterns {
		re, err := regexp.Compile(pattern)
		if err != nil {
			return nil, fmt.Errorf("invalid filename prefix pattern %q: %w", pattern, err)
		}
		filenamePrefixPatterns = append(filenamePrefixPatterns, re)
	}

	// Add default patterns
	if len(filenamePrefixPatterns) == 0 {
		defaultPatterns := []string{
			`(?i)^` + regexp.QuoteMeta(cfg.Name) + `[_-]?`,
			`(?i)^cjenik[_-]?`,
		}
		for _, pattern := range defaultPatterns {
			re, err := regexp.Compile(pattern)
			if err != nil {
				return nil, fmt.Errorf("invalid default filename prefix pattern %q: %w", pattern, err)
			}
			filenamePrefixPatterns = append(filenamePrefixPatterns, re)
		}
	}

	var rateLimitConfig ratelimit.Config
	if cfg.RateLimitOverrides != nil {
		rateLimitConfig = ratelimit.WithOverrides(*cfg.RateLimitOverrides)
	} else {
		rateLimitConfig = ratelimit.DefaultConfig()
	}

	return &BaseChainAdapter{
		slug:                   cfg.Slug,
		name:                   cfg.Name,
		supportedTypes:         cfg.SupportedTypes,
		config:                 cfg.ChainConfig,
		filenamePrefixPatterns: filenamePrefixPatterns,
		fileExtensionPattern:   fileExtensionPattern,
		rateLimiter:            ratelimit.NewRateLimiter(rateLimitConfig),
		rateLimitConfig:        rateLimitConfig,
		httpClient:             httpclient.NewClient(rateLimitConfig),
	}, nil
}

// Slug returns the chain slug
func (a *BaseChainAdapter) Slug() string {
	return a.slug
}

// Name returns the chain name
func (a *BaseChainAdapter) Name() string {
	return a.name
}

// SupportedTypes returns supported file types
func (a *BaseChainAdapter) SupportedTypes() []types.FileType {
	return a.supportedTypes
}

// BaseURL returns the base URL for the chain's portal
func (a *BaseChainAdapter) BaseURL() string {
	return a.config.BaseURL
}

// HTTPClient returns the HTTP client for making requests
func (a *BaseChainAdapter) HTTPClient() *httpclient.Client {
	return a.httpClient
}

// Discover discovers available price files from the chain's portal
// targetDate is optional (can be empty string) - subclasses may use it for date-specific discovery
func (a *BaseChainAdapter) Discover(targetDate string) ([]types.DiscoveredFile, error) {
	baseURL := a.config.BaseURL
	discoveredFiles := make([]types.DiscoveredFile, 0)

	resp, err := a.httpClient.Get(baseURL)
	if err != nil {
		return nil, fmt.Errorf("failed to fetch %s portal: %w", a.name, err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("failed to fetch %s portal: status %d", a.name, resp.StatusCode)
	}

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, fmt.Errorf("failed to read response body: %w", err)
	}

	html := string(body)

	// Get file extensions to look for
	extensions := a.getDiscoverableExtensions()
	extensionPattern := strings.Join(extensions, "|")
	linkPattern := regexp.MustCompile(`(?i)href=["']([^"']*\.(` + extensionPattern + `)(?:\?[^"']*)?)["']`)

	matches := linkPattern.FindAllStringSubmatch(html, -1)
	seenURLs := make(map[string]bool)

	for _, match := range matches {
		if len(match) < 2 {
			continue
		}

		href := match[1]
		if seenURLs[href] {
			continue
		}
		seenURLs[href] = true

		var fileURL string
		if strings.HasPrefix(href, "http://") || strings.HasPrefix(href, "https://") {
			fileURL = href
		} else {
			parsedURL, err := url.Parse(baseURL)
			if err != nil {
				continue
			}
			if strings.HasPrefix(href, "/") {
				fileURL = fmt.Sprintf("%s://%s%s", parsedURL.Scheme, parsedURL.Host, href)
			} else {
				basePath := parsedURL.Path
				if idx := strings.LastIndex(basePath, "/"); idx >= 0 {
					fileURL = fmt.Sprintf("%s://%s%s%s", parsedURL.Scheme, parsedURL.Host, basePath[:idx+1], href)
				} else {
					fileURL = fmt.Sprintf("%s://%s/%s", parsedURL.Scheme, parsedURL.Host, href)
				}
			}
		}

		filename := a.extractFilenameFromURL(fileURL)
		fileType := a.detectFileType(filename)

		discoveredFiles = append(discoveredFiles, types.DiscoveredFile{
			URL:      fileURL,
			Filename: filename,
			Type:     fileType,
			Size:     nil,
			LastModified: nil,
			Metadata: map[string]string{
				"source":       fmt.Sprintf("%s_portal", a.slug),
				"discoveredAt": time.Now().Format(time.RFC3339),
			},
		})
	}

	return discoveredFiles, nil
}

// getDiscoverableExtensions returns file extensions to look for during discovery
func (a *BaseChainAdapter) getDiscoverableExtensions() []string {
	extensions := make([]string, 0)
	for _, fileType := range a.supportedTypes {
		switch fileType {
		case types.FileTypeCSV:
			extensions = append(extensions, "csv")
		case types.FileTypeXLSX:
			extensions = append(extensions, "xlsx", "xls")
		case types.FileTypeXML:
			extensions = append(extensions, "xml")
		case types.FileTypeZIP:
			extensions = append(extensions, "zip")
		}
	}
	return extensions
}

// extractFilenameFromURL extracts filename from URL
func (a *BaseChainAdapter) extractFilenameFromURL(urlStr string) string {
	parsedURL, err := url.Parse(urlStr)
	if err != nil {
		return fmt.Sprintf("unknown.%s", a.supportedTypes[0])
	}

	pathname := parsedURL.Path
	parts := strings.Split(pathname, "/")
	filename := parts[len(parts)-1]

	if filename == "" {
		return fmt.Sprintf("unknown.%s", a.supportedTypes[0])
	}

	return strings.Split(filename, "?")[0]
}

// detectFileType detects file type from filename
func (a *BaseChainAdapter) detectFileType(filename string) types.FileType {
	lowerFilename := strings.ToLower(filename)
	if strings.HasSuffix(lowerFilename, ".csv") {
		return types.FileTypeCSV
	}
	if strings.HasSuffix(lowerFilename, ".xlsx") || strings.HasSuffix(lowerFilename, ".xls") {
		return types.FileTypeXLSX
	}
	if strings.HasSuffix(lowerFilename, ".xml") {
		return types.FileTypeXML
	}
	if strings.HasSuffix(lowerFilename, ".zip") {
		return types.FileTypeZIP
	}
	return a.supportedTypes[0]
}

// Fetch fetches a discovered file with rate limiting and retry logic
func (a *BaseChainAdapter) Fetch(file types.DiscoveredFile) (*types.FetchedFile, error) {
	content, err := a.httpClient.GetBytes(file.URL)
	if err != nil {
		return nil, fmt.Errorf("failed to fetch file: %w", err)
	}

	hash := httpclient.ComputeSha256(content)

	return &types.FetchedFile{
		Discovered: file,
		Content:    content,
		Hash:       hash,
	}, nil
}

// Parse must be implemented by subclasses
func (a *BaseChainAdapter) Parse(content []byte, filename string, options *types.ParseOptions) (*types.ParseResult, error) {
	return nil, fmt.Errorf("Parse method must be implemented by subclass")
}

// ExtractStoreIdentifier extracts store identifier from discovered file
func (a *BaseChainAdapter) ExtractStoreIdentifier(file types.DiscoveredFile) *types.StoreIdentifier {
	identifier := a.extractStoreIdentifierFromFilename(file.Filename)
	if identifier == "" {
		return nil
	}

	return &types.StoreIdentifier{
		Type:  "filename_code",
		Value: identifier,
	}
}

// extractStoreIdentifierFromFilename extracts store identifier string from filename
func (a *BaseChainAdapter) extractStoreIdentifierFromFilename(filename string) string {
	baseName := a.fileExtensionPattern.ReplaceAllString(filename, "")

	cleanName := baseName
	for _, pattern := range a.filenamePrefixPatterns {
		cleanName = pattern.ReplaceAllString(cleanName, "")
	}

	cleanName = strings.TrimSpace(cleanName)

	if cleanName == "" {
		return a.fileExtensionPattern.ReplaceAllString(filename, "")
	}

	return cleanName
}

// ValidateRow validates a normalized row
func (a *BaseChainAdapter) ValidateRow(row types.NormalizedRow) types.NormalizedRowValidation {
	var errors, warnings []string

	if strings.TrimSpace(row.Name) == "" {
		errors = append(errors, "Missing product name")
	}

	if row.Price <= 0 {
		errors = append(errors, "Price must be positive")
	}

	if row.Price > 100000000 {
		warnings = append(warnings, "Price seems unusually high")
	}

	if row.DiscountPrice != nil && *row.DiscountPrice >= row.Price {
		warnings = append(warnings, "Discount price is not less than regular price")
	}

	for _, barcode := range row.Barcodes {
		if !isValidBarcode(barcode) {
			warnings = append(warnings, fmt.Sprintf("Invalid barcode format: %s", barcode))
		}
	}

	return types.NormalizedRowValidation{
		IsValid:  len(errors) == 0,
		Errors:   errors,
		Warnings: warnings,
	}
}

// isValidBarcode checks if barcode is valid format
func isValidBarcode(barcode string) bool {
	if len(barcode) < 8 || len(barcode) > 14 {
		return false
	}
	for _, c := range barcode {
		if c < '0' || c > '9' {
			return false
		}
	}
	return true
}

// ExtractStoreMetadata extracts store metadata from file for auto-registration
func (a *BaseChainAdapter) ExtractStoreMetadata(file types.DiscoveredFile) *types.StoreMetadata {
	identifier := a.extractStoreIdentifierFromFilename(file.Filename)
	if identifier == "" {
		return nil
	}

	return &types.StoreMetadata{
		Name: fmt.Sprintf("%s %s", a.name, identifier),
	}
}

File: src/ingestion/adapters/base/chain.ts
import { createHash } from "node:crypto";
import type { Result } from "neverthrow";
import { err, errAsync, ok, okAsync, ResultAsync } from "neverthrow";
import { type FetchError, fetchError } from "@/lib/errors";
import type { IngestionClassified } from "../../errors";
import type {
	DiscoveredFile,
	FetchedFile,
	FileType,
	NormalizedRow,
	NormalizedRowValidation,
	ParseOptions,
	ParseResult,
	StoreIdentifier,
	StoreMetadata,
} from "../../types";
import type { ChainConfig } from "../config";

export interface RateLimitConfig {
	requestsPerSecond: number;
	maxRetries: number;
	initialBackoffMs: number;
	maxBackoffMs: number;
}

export interface PartialRateLimitConfig {
	requestsPerSecond?: number;
	maxRetries?: number;
	initialBackoffMs?: number;
	maxBackoffMs?: number;
}

export interface BaseAdapterConfig {
	slug: string;
	name: string;
	supportedTypes: FileType[];
	chainConfig: ChainConfig;
	filenamePrefixPatterns?: string[];
	fileExtensionPattern?: RegExp;
	rateLimitOverrides?: PartialRateLimitConfig;
}

const defaultRateLimitConfig: RateLimitConfig = {
	requestsPerSecond: 2,
	maxRetries: 3,
	initialBackoffMs: 100,
	maxBackoffMs: 30_000,
};

class RateLimiter {
	private config: RateLimitConfig;
	private lastRequest: number;

	constructor(config: RateLimitConfig) {
		this.config = config;
		this.lastRequest = 0;
	}

	getConfig(): RateLimitConfig {
		return this.config;
	}

	async throttle(): Promise<void> {
		const now = Date.now();
		const minInterval = 1000 / this.config.requestsPerSecond;
		const elapsed = now - this.lastRequest;
		if (elapsed < minInterval) {
			await sleep(minInterval - elapsed);
		}
		this.lastRequest = Date.now();
	}
}

export class BaseChainAdapter {
	slug: string;
	name: string;
	supportedTypes: FileType[];
	protected config: ChainConfig;
	protected filenamePrefixPatterns: RegExp[];
	protected fileExtensionPattern: RegExp;
	protected rateLimiter: RateLimiter;
	protected rateLimitConfig: RateLimitConfig;
	protected initializationError: FetchError | null;

	constructor(cfg: BaseAdapterConfig) {
		const configuredTypes = cfg.supportedTypes.length
			? cfg.supportedTypes
			: cfg.chainConfig.supportedTypes;
		this.initializationError =
			configuredTypes.length === 0
				? fetchError({
						url: cfg.chainConfig.baseUrl,
						message: `${cfg.slug}: supportedTypes cannot be empty`,
						retryable: false,
						attempts: 0,
					})
				: null;

		this.slug = cfg.slug;
		this.name = cfg.name;
		this.supportedTypes = configuredTypes.length > 0 ? configuredTypes : ["csv"];
		this.config = cfg.chainConfig;

		this.fileExtensionPattern = cfg.fileExtensionPattern ?? /\.(csv|CSV)$/;

		const prefixPatterns = cfg.filenamePrefixPatterns ?? [];
		this.filenamePrefixPatterns = prefixPatterns
			.map((pattern) => compilePattern(pattern))
			.concat(
				prefixPatterns.length === 0
					? [
							new RegExp(`^${escapeRegex(this.name)}[_-]?`, "i"),
							/^cjenik[_-]?/i,
						]
					: [],
			);

		this.rateLimitConfig = {
			...defaultRateLimitConfig,
			...cfg.rateLimitOverrides,
		};
		this.rateLimiter = new RateLimiter(this.rateLimitConfig);
	}

	baseUrl(): string {
		return this.config.baseUrl;
	}

	discover(
		_targetDate?: string,
	): ResultAsync<DiscoveredFile[], FetchError | IngestionClassified> {
		if (this.initializationError) {
			return errAsync(this.initializationError);
		}

		return this.fetchWithRetry(this.config.baseUrl)
			.andThen((response) =>
				ResultAsync.fromPromise(response.text(), (e) =>
					fetchError({
						url: this.config.baseUrl,
						message: e instanceof Error ? e.message : "Failed to read response",
						retryable: false,
						attempts: 1,
						cause: e,
					}),
				),
			)
			.andThen((html) => {
				const extensions = this.getDiscoverableExtensions();
				const extensionPattern = extensions.join("|");
				const linkPattern = new RegExp(
					`href=["']([^"']*\\.(?:${extensionPattern})(?:\\?[^"']*)?)["']`,
					"gi",
				);

				const matches = html.matchAll(linkPattern);
				const seen = new Set<string>();
				const files: DiscoveredFile[] = [];

				for (const match of matches) {
					const href = match[1];
					if (!href || seen.has(href)) {
						continue;
					}
					seen.add(href);

					const fileUrl = resolveUrl(this.config.baseUrl, href);
					const filename = this.extractFilenameFromUrl(fileUrl);
					const type = this.detectFileType(filename);

					files.push({
						url: fileUrl,
						filename,
						type,
						metadata: {
							source: `${this.slug}_portal`,
							discoveredAt: new Date().toISOString(),
						},
					});
				}

				return okAsync(files);
			});
	}

	fetch(file: DiscoveredFile): ResultAsync<FetchedFile, FetchError> {
		if (this.initializationError) {
			return errAsync(this.initializationError);
		}

		return this.fetchWithRetry(file.url)
			.andThen((response) =>
				ResultAsync.fromPromise(response.arrayBuffer(), (e) =>
					fetchError({
						url: file.url,
						message:
							e instanceof Error ? e.message : "Failed to read response body",
						retryable: false,
						attempts: 1,
						cause: e,
					}),
				),
			)
			.map((arrayBuffer) => {
				const buffer = Buffer.from(arrayBuffer);
				return {
					discovered: file,
					content: buffer,
					hash: computeSha256(buffer),
				};
			});
	}

	parse(
		_content: Buffer,
		_filename: string,
		_options?: ParseOptions,
	): ResultAsync<ParseResult, FetchError> {
		if (this.initializationError) {
			return errAsync(this.initializationError);
		}

		return errAsync(
			fetchError({
				url: "",
				message: "Parse method must be implemented by subclass",
				retryable: false,
				attempts: 0,
			}),
		);
	}

	extractStoreIdentifier(file: DiscoveredFile): StoreIdentifier | null {
		const identifier = this.extractStoreIdentifierFromFilename(file.filename);
		if (!identifier) {
			return null;
		}
		return {
			type: "filename_code",
			value: identifier,
		};
	}

	validateRow(row: NormalizedRow): NormalizedRowValidation {
		const errors: string[] = [];
		const warnings: string[] = [];

		if (!row.name || row.name.trim() === "") {
			errors.push("Missing product name");
		}

		if (row.priceStatus === "available") {
			if (row.price === null || row.price <= 0) {
				errors.push("Price must be positive when marked available");
			}
		} else if (row.price !== null && row.price > 0) {
			warnings.push("Unavailable row has a positive price value");
		}

		if (row.price !== null && row.price > 100_000_000) {
			warnings.push("Price seems unusually high");
		}

		if (
			row.price !== null &&
			row.discountPrice !== undefined &&
			row.discountPrice >= row.price
		) {
			warnings.push("Discount price is not less than regular price");
		}

		for (const barcode of row.barcodes) {
			if (!isValidBarcode(barcode)) {
				warnings.push(`Invalid barcode format: ${barcode}`);
			}
		}

		return {
			isValid: errors.length === 0,
			errors,
			warnings,
		};
	}

	extractStoreMetadata(file: DiscoveredFile): StoreMetadata | null {
		const identifier = this.extractStoreIdentifierFromFilename(file.filename);
		if (!identifier) {
			return null;
		}
		return {
			name: `${this.name} ${identifier}`,
		};
	}

	protected extractFilenameFromUrl(url: string): string {
		try {
			const parsed = new URL(url);
			const pathname = parsed.pathname;
			const parts = pathname.split("/");
			const filename = parts[parts.length - 1];
			return filename
				? filename.split("?")[0]
				: `unknown.${this.supportedTypes[0]}`;
		} catch {
			return `unknown.${this.supportedTypes[0]}`;
		}
	}

	protected detectFileType(filename: string): FileType {
		const lower = filename.toLowerCase();
		if (lower.endsWith(".csv")) return "csv";
		if (lower.endsWith(".xlsx") || lower.endsWith(".xls")) return "xlsx";
		if (lower.endsWith(".xml")) return "xml";
		if (lower.endsWith(".zip")) return "zip";
		return this.supportedTypes[0];
	}

	protected getDiscoverableExtensions(): string[] {
		const extensions: string[] = [];
		for (const type of this.supportedTypes) {
			switch (type) {
				case "csv":
					extensions.push("csv");
					break;
				case "xlsx":
					extensions.push("xlsx", "xls");
					break;
				case "xml":
					extensions.push("xml");
					break;
				case "zip":
					extensions.push("zip");
					break;
			}
		}
		return extensions;
	}

	protected extractStoreIdentifierFromFilename(filename: string): string {
		const baseName = filename.replace(this.fileExtensionPattern, "");
		let cleanName = baseName;
		for (const pattern of this.filenamePrefixPatterns) {
			cleanName = cleanName.replace(pattern, "");
		}
		cleanName = cleanName.trim();
		if (!cleanName) {
			return filename.replace(this.fileExtensionPattern, "");
		}
		return cleanName;
	}

	protected fetchWithRetry(url: string): ResultAsync<Response, FetchError> {
		return new ResultAsync(this._fetchWithRetryImpl(url));
	}

	private async _fetchWithRetryImpl(
		url: string,
	): Promise<Result<Response, FetchError>> {
		let lastStatus = 0;
		let lastErrorMsg = "";

		for (
			let attempt = 0;
			attempt <= this.rateLimitConfig.maxRetries;
			attempt += 1
		) {
			await this.rateLimiter.throttle();

			try {
				const response = await fetch(url, {
					method: "GET",
					headers: {
						"User-Agent": "Kosarica-Ingestion/1.0",
						Accept: "*/*",
					},
				});

				lastStatus = response.status;
				if (response.ok) {
					return ok(response);
				}

				if (
					!isRetryableStatus(response.status) ||
					attempt === this.rateLimitConfig.maxRetries
				) {
					return err(
						fetchError({
							url,
							status: response.status,
							message: `HTTP ${response.status}: ${response.statusText}`,
							retryable: isRetryableStatus(response.status),
							attempts: attempt + 1,
						}),
					);
				}

				const retryAfter = response.headers.get("Retry-After") ?? undefined;
				const delay =
					response.status === 429
						? calculateRateLimitBackoff(
								attempt,
								this.rateLimitConfig,
								retryAfter,
							)
						: calculateBackoff(attempt, this.rateLimitConfig);
				await sleep(delay);
			} catch (error) {
				lastErrorMsg = error instanceof Error ? error.message : String(error);
				if (attempt === this.rateLimitConfig.maxRetries) {
					break;
				}
				const delay = calculateBackoff(attempt, this.rateLimitConfig);
				await sleep(delay);
			}
		}

		return err(
			fetchError({
				url,
				status: lastStatus || undefined,
				message:
					`Failed to fetch ${url} after ${this.rateLimitConfig.maxRetries + 1} attempts` +
					(lastStatus ? ` (HTTP ${lastStatus})` : "") +
					(lastErrorMsg ? `: ${lastErrorMsg}` : ""),
				retryable: true,
				attempts: this.rateLimitConfig.maxRetries + 1,
			}),
		);
	}
}

function escapeRegex(value: string): string {
	return value.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
}

function compilePattern(pattern: string): RegExp {
	const trimmed = pattern.trim();
	if (trimmed.startsWith("(?i)")) {
		const cleaned = trimmed.replace("(?i)", "");
		return new RegExp(cleaned, "i");
	}
	return new RegExp(trimmed);
}

function computeSha256(content: Buffer): string {
	return createHash("sha256").update(content).digest("hex");
}

function isValidBarcode(barcode: string): boolean {
	if (barcode.length < 8 || barcode.length > 14) {
		return false;
	}
	return /^[0-9]+$/.test(barcode);
}

function isRetryableStatus(status: number): boolean {
	return status === 429 || status >= 500;
}

function calculateBackoff(attempt: number, config: RateLimitConfig): number {
	const exponential = config.initialBackoffMs * 2 ** attempt;
	const capped = Math.min(exponential, config.maxBackoffMs);
	return capped + Math.random() * 0.25 * capped;
}

function calculateRateLimitBackoff(
	attempt: number,
	config: RateLimitConfig,
	retryAfter?: string,
): number {
	if (retryAfter) {
		const seconds = Number.parseInt(retryAfter, 10);
		if (!Number.isNaN(seconds) && seconds > 0) {
			return seconds * 1000 + Math.random() * 1000;
		}
	}
	const exponential = config.initialBackoffMs * 3 ** attempt;
	const capped = Math.min(exponential, config.maxBackoffMs);
	return capped + Math.random() * 0.25 * capped;
}

function resolveUrl(baseUrl: string, href: string): string {
	if (href.startsWith("http://") || href.startsWith("https://")) {
		return href;
	}
	try {
		const base = new URL(baseUrl);
		if (href.startsWith("/")) {
			return `${base.protocol}//${base.host}${href}`;
		}
		const basePath = base.pathname;
		const prefix = basePath.includes("/")
			? basePath.slice(0, basePath.lastIndexOf("/") + 1)
			: "/";
		return `${base.protocol}//${base.host}${prefix}${href}`;
	} catch {
		return href;
	}
}

function sleep(ms: number): Promise<void> {
	return new Promise((resolve) => setTimeout(resolve, ms));
}
